src/main/java/io/vertx/core/streams/impl/OutboundWriteQueue.java

```diff
@@ -0,0 +1,361 @@
+        return DRAIN_REQUIRED_MASK | (wip.get() == highWaterMark ? QUEUE_UNWRITABLE_MASK : 0);
+      }
+      if (consume(1) == 0) {
+        return 0;
+      }
+      return drainLoop();
+    } else {
+      queue.add(element);
+      return wip.incrementAndGet() == highWaterMark ? QUEUE_UNWRITABLE_MASK : 0;
+    }
+  }
+
+  /**
+   * Let a producer thread submit an {@code element} to the queue, no delivery is attempted. Submitting an element
+   * might acquire the ownership of the queue, when it happens a {@link #drain} operation should be called
+   * to drain the queue and release the ownership.
+   *
+   * @param element the element to submit
+   * @return a bitset of [{@link {@link #QUEUE_UNWRITABLE_MASK }, {@link #DRAIN_REQUIRED_MASK }}] flags
+   */
+  public int submit(E element) {
+    queue.add(element);
+    // winning this => overflow == null
+    int pending = wip.incrementAndGet();
+    if (pending == highWaterMark) {
+      hook2();
+    }
+    return (pending == highWaterMark ? QUEUE_UNWRITABLE_MASK : 0) + ((pending == 1) ? DRAIN_REQUIRED_MASK : 0);
+  }
+
+  protected void hook2() {
+
+  }
+
+  /**
+   * Let the consumer thread drain the queue until it becomes not writable or empty, this requires
+   * the ownership of the queue acquired by {@link #DRAIN_REQUIRED_MASK} flag.
+   *
+   * @return a bitset of [{@link #DRAIN_REQUIRED_MASK}, {@link #QUEUE_WRITABLE_MASK}] flags
+   */
+  public int drain() {
+    E elt = overflow;
+    if (elt != null) {
+      if (!consumer.test(overflow)) {
+        return DRAIN_REQUIRED_MASK;
+      }
+      overflow = null;
+      if (consume(1) == 0) {
+        return 0;
+      }
+    }
+    hook();
+    return drainLoop();
+  }
+
+  /**
+   * The main drain loop, entering this loop requires a few conditions:
+   * <ul>
+   *   <li>{@link #overflow} must be {@code null}, the consumer still accepts elements</li>
+   *   <li>{@link #wip} is greater than zero (ownership of the queue)</li>
+   *   <li>only the consumer thread can execute it</li>
+   * </ul>
+   *
+   * The loop drains elements from the queue until
+   *
+   * <ul>
+   *   <li>the queue is empty (wip == 0) which releases the queue ownership</li>
+   *   <li>the {@link #consumer} rejects an element</li>
+   * </ul>
+   *
+   * When the {@link #consumer} rejects an element, the rejected element is parked
+   * in the {@link #overflow} field and the queue ownership is not released. At this point
+   * the {@link #drain()} shall be called to try again to drain the queue.
+   *
+   * @return a bitset of [{@link {@link #QUEUE_WRITABLE_MASK }, {@link #CONSUMER_PAUSED_MASK }}] flags
+   */
+  // Note : we can optimize this by passing pending as argument of this method to avoid the initial
+  private int drainLoop() {
+    int pending = wip.get();
+    if (pending == 0) {
+      throw new IllegalStateException();
+    }
+    do {
+      int consumed;
+      for (consumed = 0; consumed < pending; consumed++) {
+        E elt = queue.poll();
```

Comment:
peek, given that JCTools is pretty efficient for single consumer, is still very effective. I would suggest to save overflow to exist/being handled and perform a 2 phase op: peek (with consume) + poll.
Or split the consumption in a predicated and consumption too (still using peek/poll)

Reply:
can you suggest a change ?